### PR TITLE
Continous Benchmarking Action

### PR DESCRIPTION
**Pull request by JuanPedroGHM**
## Description

Enables performance tracking of HeAT by using [Action for Continuous Benchmarking](https://github.com/benchmark-action/github-action-benchmark)

Issue/s resolved: #

## Changes proposed:
- Moved old benchmarks to _benchmarks/old_
- Created new benchmarks meant to run on the action inside of _benchmarks/cb_
- Added **perun** as an optional dependency
- CB action runs on every push of a pull requests with the **PR talk** tag
- Introduced dedicated **create_spherical_dataset** method in the utils module that can be used in unit tests and benchmarks

## Type of change
- New Github workflow for pull requests added.
- New benchmarks organization

## Performance
- perun works with MPI, and won't print duplicate results when running on multiple processes. 

## Due Diligence
- [ ] All split configurations tested
- [ ] Multiple dtypes tested in relevant functions
- [ ] Documentation updated (if needed)
- [X] Title of PR is suitable for corresponding CHANGELOG entry

#### Does this change modify the behaviour of other functions? If so, which?
no

### CODE REVIEW

**Review by mrfh92 — APPROVED** [review]
**Review for PR #1137 "Continuous Benchmarking Action" 

This PR adds the possibility of continuous performance monitoring to Heat and therefore it is a highly valuable contribution to the overall functionality. 

As far as I understand, there are three main changes coming along with this PR: First, those things required for the actual continuous benchmarking (including examples covering clustering and linear algebra); second, the "old" benchmarks have been moved to make place for the "new" continuous benchmarks; third, the unit tests for clustering have been cleaned up by introducing a dedicated function for creating the spherical test data set that now can also be used for benchmarking purposes. 

As far as I can judge, the code is correct. Since the CI runs through, I recommend merging.

**Comment by JuanPedroGHM** [comment]
run bench

File: .github/workflows/benchmark_pr.yml [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v3
+      - name: Setup MPI
+        uses: mpi4py/setup-mpi@v1
+      - name: Use Python 3.8
+        uses: actions/setup-python@v4
+        with:
+          python-version: 3.8 # Perun only supports 3.8 and ahead
```
Comment (ClaudiaComito) [inline .github/workflows/benchmark_pr.yml:20]: 3.9 or 3.10?

File: .github/workflows/benchmark_pr.yml [hunk 2]
```diff
@@ -0,0 +1,45 @@
+name: Benchmark PR
+on:
+  pull_request:
+    types: [opened, synchronize, reopened, labeled]
+    branches: [main]
+
+jobs:
+  benchmark-pr:
+    name: Benchmark PR
+    if: contains(github.event.pull_request.labels.*.name, 'PR talk')
```
Comment (ClaudiaComito) [inline .github/workflows/benchmark_pr.yml:10]: Let's have a dedicated label, "benchmark PR" or something? Would make it more intuitive esp. for future contributors.

File: .github/workflows/benchmark_main.yml [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+jobs:
+  benchmark-main:
+    name: Benchmark main and save
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v3
+      - name: Setup MPI
+        uses: mpi4py/setup-mpi@v1
+      - name: Use Python 3.8
```
[the remark this answers is not in the document]
Comment (JuanPedroGHM) [reply .github/workflows/benchmark_main.yml:16]: I'll change both of them to python 3.10

File: .github/workflows/benchmark_pr.yml (hunk 2 above)
Comment (JuanPedroGHM) [reply .github/workflows/benchmark_pr.yml:10]: Good idea!

**Review by ClaudiaComito — APPROVED** [review]
Wonderful, thanks a lot @JuanPedroGHM !

**Pull request merged**